### PR TITLE
Fishingmap/encounter events

### PR DESCRIPTION
**Pull request by j8seangel**
image

- [x] Show events section
- [x] New dataview category
- [x] Layer composer tile cluster generator 
- [x] Move mocked dataview to API
- [x] Pair info with related datasets
- [x] Create dataset in API
  - [x] new type
  - [x] associate with the above dataview
- [x] Add cluster click interaction
- [x] Add event click interaction, 
  - [x] grab event_id to fetch event info
  - [x] Fetch vessel info details by id
  - [x] Style popup and link to CVP
- [x] Timerange filtering
- [x] Fetch latest carrier dataset available to cvp link
- [x] Disable cvp link when event is out of the latest dataset range
- [x] ~Review styles and brainstorm about sizes~ Not for now

Not related, but broken:
- [x] Fix initial viewport on first load

### CODE REVIEW

File: applications/fishing-map/src/features/map/map.slice.ts [hunk 1]
```diff
@@ -1,34 +1,58 @@
...
+export type ExtendedFeatureEvent = ApiEvent & { dataset: Dataset }
+
+export type SliceExtendedFeature = ExtendedFeature & {
+  event?: ExtendedFeatureEvent
+  vessels?: ExtendedFeatureVessel[]
+}
+
+// Extends the default extendedEvent including event and vessels information from API
+export type SliceInteractionEvent = Omit<InteractionEvent, 'features'> & {
+  features: SliceExtendedFeature[]
```
Comment (j8seangel) [inline applications/fishing-map/src/features/map/map.slice.ts:41]: Comes from ExtendedFeature as the vessels list and events are app related instead of the hook responsability

File: applications/fishing-map/src/features/map/map.slice.ts [hunk 2]
```diff
@@ -200,12 +224,39 @@ export const fetch4WingInteractionThunk = createAsyncThunk<
...
+  const state = getState() as RootState
+  const eventDataviews = selectEventsDataviews(state) || []
+  const dataview = eventDataviews.find((d) => d.id === eventFeature.generatorId)
+  const dataset = dataview?.datasets?.find((d) => d.type === DatasetTypes.Events)
+  if (dataset) {
+    // workaround using another dataset from v0 until the events API is ready to fetch events by id
+    const eventDataset = getRelatedDatasetByType(dataset, DatasetTypes.Events)
+    if (eventDataset) {
+      // TODO use v1 dataset endpoint instead of hardcoded v0 when ready
+      const eventUrl = `https://gateway.api.dev.globalfishingwatch.org/datasets/${eventDataset.id}/events/${eventFeature.id}`
```
Comment (j8seangel) [inline applications/fishing-map/src/features/map/map.slice.ts:243]: TODO asap, remove this

File: applications/fishing-map/src/features/map/popups/TileClusterLayers.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+          ?.id
+        const linkParams = {
+          ...viewport,
+          dataset: eventDataset,
+          vessel: feature.event?.vessel?.id,
+          ...(feature.event && { timestamp: new Date(feature.event.start).getTime() }),
+          start,
+          end,
+        }
+        const urlLink = `${CARRIER_PORTAL_URL}/?${stringify(linkParams)}`
```
Comment (j8seangel) [inline applications/fishing-map/src/features/map/popups/TileClusterLayers.tsx:64]: Remember to include the new .env to get this working, see example in `.env.sample`

File: applications/fishing-map/src/features/workspace/workspace.slice.ts [hunk 4]
```diff
@@ -94,29 +94,29 @@ export const fetchWorkspaceThunk = createAsyncThunk(
           dataviews = payload
         }
       }
-      const dataviewIntances = [
+      const dataviewInstances = [
```
Comment (j8seangel) [inline applications/fishing-map/src/features/workspace/workspace.slice.ts:97]: this just fixes a typo

File: applications/fishing-map/src/index.tsx [hunk 5]
```diff
@@ -27,8 +26,3 @@ render(
   </React.StrictMode>,
   document.getElementById('root')
 )
-
-// If you want your app to work offline and load faster, you can change
-// unregister() to register() below. Note this comes with some pitfalls.
-// Learn more about service workers: https://bit.ly/CRA-PWA
-serviceWorker.unregister()
```
Comment (j8seangel) [inline applications/fishing-map/src/index.tsx:34]: Removed as it isn't used

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  _getStyleLayers = (config: GlobalTileClusterGeneratorConfig): AnyLayer[] => {
+    const layers = [
+      {
+        id: 'clusters',
+        type: 'circle',
+        source: config.id,
+        'source-layer': 'points',
+        filter: ['>', ['get', 'count'], 1],
+        paint: {
+          'circle-radius': [
```
Comment (j8seangel) [inline packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:55]: To discuss how to better calculate this sizes

**Review by j8seangel — COMMENTED** [review]
Some comments to make this review a bit easier

File: applications/fishing-map/src/features/app/App.tsx [hunk 7]
```diff
@@ -108,8 +111,11 @@ function App(): React.ReactElement {
   useEffect(() => {
     const fetchWorkspace = async () => {
       const action = await dispatch(fetchWorkspaceThunk(urlWorkspaceId as string))
-      if (fetchWorkspaceThunk.fulfilled.match(action) && action.payload?.viewport) {
-        setMapCoordinates(action.payload.viewport)
+      if (fetchWorkspaceThunk.fulfilled.match(action)) {
```
Comment (j8seangel) [inline applications/fishing-map/src/features/app/App.tsx:114]: Fixes first viewport position, we need to wait until workspace is loaded to use url or workspace params

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  type = Type.TileCluster
+
+  _getStyleSources = (config: GlobalTileClusterGeneratorConfig) => {
+    if (!config.tilesUrl) {
+      throw new Error(`Tile Cluster layer should specify tilesUrl ${JSON.stringify(config)}`)
+    }
+    const tilesUrl = isUrlAbsolute(config.tilesUrl)
+      ? config.tilesUrl
+      : API_GATEWAY + config.tilesUrl
+    const url = new URL(tilesUrl.replace(/{{/g, '{').replace(/}}/g, '}'))
```
Comment (nerik) [inline packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:22]: Could this be fixed in the API instead?

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts (hunk 6 above)
Comment (nerik) [reply packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:55]: Always use square root when mapping a variable to the circle surface - which is what you really want

Screenshot 2021-03-24 at 13 05 12

https://blog.mapbox.com/announcing-expressions-in-gl-js-a72b55d0a6af

Then I suppose an exponential scale on top of that makes sense.

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      {
+        id: 'cluster_count',
+        type: 'symbol',
+        source: config.id,
+        'source-layer': 'points',
+        filter: ['>', ['get', 'count'], 1],
+        layout: {
+          'text-size': 14,
+          'text-field': ['get', 'count'],
+          'text-font': ['Roboto Mono Light'],
```
Comment (nerik) [inline packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:86]: @satellitestudiodesign Why using Roboto Mono? As far as I know it's the only place where we use it no? Seems a little bit off to me

Screenshot 2021-03-24 at 11 48 46

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      {
+        id: 'unclustered-point',
+        type: 'circle',
+        source: config.id,
+        'source-layer': 'points',
+        filter: ['==', ['get', 'count'], 1],
+        layout: {
+          visibility: isConfigVisible(config),
+        },
+        paint: {
```
Comment (nerik) [inline packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:106]: Lo del icono

File: packages/layer-composer/src/generators/types.ts [hunk 11]
```diff
@@ -131,6 +132,30 @@ export interface ContextGeneratorConfig extends GeneratorConfig {
...
+   */
+  color?: string
+  /**
+   * Url to grab the tiles from, internally using https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/#vector-tiles
+   */
+  tilesUrl: string
+  /**
+   * List of types supported by the API, optional to allow using resolved url params directly (eg: using resolve-endpoints)
+   */
+  eventTypes?: TilClusterEventType | TilClusterEventType[]
```
Comment (nerik) [inline packages/layer-composer/src/generators/types.ts:152]:
```suggestion
  eventTypes?: TileClusterEventType | TileClusterEventType[]
```

**Review by nerik — COMMENTED** [review]
Happy to see this feature elegantly implemented knowing the porqueria that I made in v2 🥳 
cc @satellitestudiodesign 

Some comments/questions about functionality/design:
- encounters icon is missing in non-clustered points. Would be nice to be inline visually with the CVP (the icon should already be available in our MGL spritesheet).
- should we follow a convention for track colors when pinning them like in v2? (carrier in a color by default, donor with another)
- Not sure if bug or feature or openstreetmap or escoffier:
Screenshot 2021-03-24 at 11 45 32
'See in carrier' button is grayed out with 🚫 cursor but it actually works

- Basically what's up with this workspace? We should actually enable it no?
Screenshot 2021-03-24 at 11 48 21

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts (hunk 9 above)
Comment (satellitestudiodesign) [reply packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:86]: I'd change it to Roboto Medium, yes. Also available in the `map-gl-glyphs` repo

File: applications/fishing-map/src/features/map/map.hooks.ts [hunk 12]
```diff
@@ -86,8 +97,30 @@ export const useClickedEventConnect = () => {
...
 
-    if (promiseRef.current) {
-      promiseRef.current.abort()
+    const clusterFeature = event?.features?.find(
+      (f) => f.generatorType === Generators.Type.TileCluster
+    )
+    if (clusterFeature?.properties?.expansionZoom) {
+      const { count, expansionZoom, lat, lng } = clusterFeature.properties
+      if (count > 1) {
+        if (!tilesLoading) {
```
[the remark this answers is not in the document]
Comment (j8seangel) [reply applications/fishing-map/src/features/map/map.hooks.ts:106]: I've replaced the general tilesLoading by
```js
const encounterSourceLoading = useMapSourceLoaded(TILE_CLUSTER_SOURCE_ID)
```
to set the loading only for that source

File: applications/fishing-map/src/features/map/map.slice.ts [hunk 13]
```diff
@@ -1,34 +1,85 @@
...
 export const MAX_TOOLTIP_VESSELS = 5
 
+export type ExtendedFeatureVessel = {
+  id: string
+  hours: number
+  dataset: Dataset
+  [key: string]: any
+}
+
+type TempClusterBusterEvent = {
```
[the remark this answers is not in the document]
Comment (j8seangel) [reply applications/fishing-map/src/features/map/map.slice.ts:35]: Because this is out of the regular API event format... will update the name to use Temporal and a comment explaining it

File: applications/amathea/CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,34 @@
 # @globalfishingwatchapp/amathea
 
+## 0.14.9
```
[the remark this answers is not in the document]
Comment (j8seangel) [reply applications/amathea/CHANGELOG.md:3]: yes, it is automatically generated by the changesets... I'll create a branch with the project in case we need to recover something and remove it completely in another branch

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts (hunk 10 above)
Comment (j8seangel) [reply packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:106]: 🍨

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts (hunk 9 above)
Comment (j8seangel) [reply packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:86]: Done!

File: packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts (hunk 8 above)
Comment (j8seangel) [reply packages/layer-composer/src/generators/tile-cluster/tile-cluster.ts:22]: Noted and I will open the conversation in the next meeting

**Comment by j8seangel** [comment]
I couldn't manage to get the "sqrt" expression working, I'm going to merge as it is for now and let's discuss it later
